// @jup-ag/crema-sdk/dist/cjs/instructions/depositFixToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.depositFixTokenInstruction = exports.Decode = void 0;
const buffer_layout_1 = require("@solana/buffer-layout");
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
const layout_1 = require("../util/layout");
const instruction_1 = require("./instruction");
const DataLayout = (0, buffer_layout_1.struct)([
    (0, buffer_layout_1.u8)("instruction"),
    (0, buffer_layout_1.u8)("isNewPosition"),
    (0, buffer_layout_1.u8)("fixTokenType"),
    (0, buffer_layout_1.s32)("tickLower"),
    (0, buffer_layout_1.s32)("tickUpper"),
    (0, layout_1.decimalU64)("maximumTokenA"),
    (0, layout_1.decimalU64)("maximumTokenB"),
    (0, layout_1.decimalU64)("positionIndex"),
]);
const Decode = (data) => {
    return DataLayout.decode(data);
};
exports.Decode = Decode;
const depositFixTokenInstruction = (programId, tokenSwapKey, authority, userTransferAuthority, userTokenA, userTokenB, swapTokenA, swapTokenB, nftMint, nftUser, ticksKey, positionsKey, isNewPosition, fixTokenType, tickLower, tickUpper, maximumTokenA, maximumTokenB, positionIndex) => {
    const data = Buffer.alloc(DataLayout.span);
    DataLayout.encode({
        instruction: instruction_1.TokenSwapInstruction.DepositFixToken,
        isNewPosition,
        fixTokenType,
        tickLower,
        tickUpper,
        maximumTokenA,
        maximumTokenB,
        positionIndex,
    }, data);
    console.log({
        instruction: instruction_1.TokenSwapInstruction.DepositFixToken,
        isNewPosition,
        fixTokenType,
        tickLower,
        tickUpper,
        maximumTokenA: maximumTokenA.toString(),
        maximumTokenB: maximumTokenB.toString(),
        positionIndex: positionIndex.toString(),
    });
    const keys = [
        { pubkey: tokenSwapKey, isSigner: false, isWritable: true },
        { pubkey: authority, isSigner: false, isWritable: false },
        { pubkey: userTransferAuthority, isSigner: true, isWritable: false },
        { pubkey: userTokenA, isSigner: false, isWritable: true },
        { pubkey: userTokenB, isSigner: false, isWritable: true },
        { pubkey: swapTokenA, isSigner: false, isWritable: true },
        { pubkey: swapTokenB, isSigner: false, isWritable: true },
        { pubkey: nftMint, isSigner: false, isWritable: true },
        { pubkey: nftUser, isSigner: false, isWritable: true },
        { pubkey: ticksKey, isSigner: false, isWritable: true },
        { pubkey: positionsKey, isSigner: false, isWritable: true },
        { pubkey: spl_token_1.TOKEN_PROGRAM_ID, isSigner: false, isWritable: false },
    ];
    return new web3_js_1.TransactionInstruction({
        keys,
        programId,
        data,
    });
};
exports.depositFixTokenInstruction = depositFixTokenInstruction;
