// @jup-ag/crema-sdk/dist/cjs/instructions/instruction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenSwapInstruction = void 0;
/** @internal */
var TokenSwapInstruction;
(function (TokenSwapInstruction) {
    TokenSwapInstruction[TokenSwapInstruction["Initialize"] = 0] = "Initialize";
    TokenSwapInstruction[TokenSwapInstruction["Swap"] = 1] = "Swap";
    TokenSwapInstruction[TokenSwapInstruction["DepositAllTokenTypes"] = 2] = "DepositAllTokenTypes";
    TokenSwapInstruction[TokenSwapInstruction["WithdrawAllTokenTypes"] = 3] = "WithdrawAllTokenTypes";
    TokenSwapInstruction[TokenSwapInstruction["Claim"] = 4] = "Claim";
    TokenSwapInstruction[TokenSwapInstruction["ManagerClaim"] = 5] = "ManagerClaim";
    TokenSwapInstruction[TokenSwapInstruction["AddUserPosition"] = 6] = "AddUserPosition";
    TokenSwapInstruction[TokenSwapInstruction["UpdateFee"] = 7] = "UpdateFee";
    TokenSwapInstruction[TokenSwapInstruction["SimulateSwap"] = 8] = "SimulateSwap";
    TokenSwapInstruction[TokenSwapInstruction["DepositFixToken"] = 9] = "DepositFixToken";
})(TokenSwapInstruction = exports.TokenSwapInstruction || (exports.TokenSwapInstruction = {}));
